### ISSUE TITLE
ENOTFOUND hub-cloud.browserstack.com in WebdriverIO v6

### ISSUE THREAD

**Issue by th0car** [webdriverio/webdriverio#5378]
**Environment (please complete the following information):**
 - **WebdriverIO version:** 6.1.9
 - **Mode:**  WDIO Testrunner
 - **If WDIO Testrunner, running** sync
 - **Node.js version:** 12.16.2
 - **NPM version:** 6.14.4
 - **Browser name and version:** Chrome
 - **Platform name and version:** Windows 10
 - **Additional wdio packages used (if applicable):**  @wdio/cli": "^6.1.9", @wdio/local-runner": "^6.1.9", "@wdio/mocha-framework": "^6.1.8", "@wdio/sync": "^6.1.8"

**Config of WebdriverIO**
```
exports.config = {
  user: process.env.BROWSERSTACK_USERNAME || "myUsername",
  key: process.env.BROWSERSTACK_ACCESS_KEY || "myAccessKey",

  updateJob: false,
  specs: [
    './tests/specs/single_test.js'
  ],
  exclude: [],

  capabilities: [{
    browser: 'Chrome',
    name: 'single_test',
    build: 'webdriver-browserstack'
  }],
  logLevel: 'warn',
  coloredLogs: true,
  screenshotPath: './errorShots/',
  baseUrl: '',
  waitforTimeout: 10000,
  connectionRetryTimeout: 90000,
  connectionRetryCount: 3

  before: function () {
    var chai = require('chai');
    global.expect = chai.expect;
    chai.Should();
  },
  framework: 'mocha',
  mochaOpts: {
    ui: 'bdd',
    timeout: 60000
  },
}
```

**Describe the bug**

After upgrading @wdio/cli and @wdio/local-runner to 6.1.9 tests using Browserstack starts failing with this error:
```
[0-0] 2020-05-08T14:01:13.238Z ERROR webdriver: RequestError: getaddrinfo ENOTFOUND hub-cloud.browserstack.com
    at ClientRequest.<anonymous> (C:\WS\webdriverio-browserstack-master_6nonfunctional\node_modules\got\dist\source\core\index.js:764:25)
    at Object.onceWrapper (events.js:417:26)
    at ClientRequest.emit (events.js:322:22)
    at ClientRequest.EventEmitter.emit (domain.js:482:12)
    at ClientRequest.origin.emit (C:\WS\webdriverio-browserstack-master_6nonfunctional\node_modules\@szmarczak\http-timer\dist\source\index.js:39:20)
    at TLSSocket.socketErrorListener (_http_client.js:426:9)
    at TLSSocket.emit (events.js:310:20)
    at TLSSocket.EventEmitter.emit (domain.js:482:12)
    at emitErrorNT (internal/streams/destroy.js:92:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
    at GetAddrInfoReqWrap.onlookupall [as oncomplete] (dns.js:76:26)
[0-0]  Error:  Failed to create session.
getaddrinfo ENOTFOUND hub-cloud.browserstack.com
[0-0] 2020-05-08T14:01:13.238Z ERROR @wdio/runner: Error: Failed to create session.
getaddrinfo ENOTFOUND hub-cloud.browserstack.com
    at startWebDriverSession (C:\WS\webdriverio-browserstack-master_6nonfunctional\node_modules\webdriver\build\utils.js:45:11)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

**To Reproduce**

1. Install the browserstack-webdriverIO v5 example application from https://github.com/browserstack/webdriverio-browserstack
2. Update single.conf.js with valid Browserstack User and Key 
3. Run test "single" - it should pass without error and successful lookup and connection to hub-cloud.browserstack.com is made.
4. Update wedriverio and @wdio/cli and/or @wdio/local-runner to 6.1.9  The rest of the dependencies can be updated or not - it wont affect this problem. **Make no changes whatsoever to conf or spec.**
5. Run test "single"
6. Test fails with the above error.

For recreating the error a browserstack key isn't ctually needed since the error occurs before these can be validated.

Tests are executed from VS code from within a corporate network using proxy. Proxy is set in .npmrc as well as in environment varibles and is not causing any problems until upgrading WebdriverIO used for Browserstack tests.  

**Expected behavior**
host hub-cloud.browserstack.com should be resolved and test pass

**Log**
```
Execution of 1 spec files started at 2020-05-08T15:03:27.502Z

[0-0] RUNNING in undefined - C:\WS\webdriverio-browserstack-master_6nonfunctional\tests\specs\single_test.js
[0-0] (node:15368) Warning: Setting the NODE_TLS_REJECT_UNAUTHORIZED environment variable to '0' makes TLS connections and HTTPS requests insecure by disabling certificate verification.
[0-0] 2020-05-08T15:03:28.623Z ERROR webdriver: RequestError: getaddrinfo ENOTFOUND hub-cloud.browserstack.com
    at ClientRequest.<anonymous> (C:\WS\webdriverio-browserstack-master_6nonfunctional\node_modules\got\dist\source\core\index.js:764:25)
    at Object.onceWrapper (events.js:417:26)
    at ClientRequest.emit (events.js:322:22)
    at ClientRequest.EventEmitter.emit (domain.js:482:12)
    at ClientRequest.origin.emit (C:\WS\webdriverio-browserstack-master_6nonfunctional\node_modules\@szmarczak\http-timer\dist\source\index.js:39:20)
    at TLSSocket.socketErrorListener (_http_client.js:426:9)
    at TLSSocket.emit (events.js:310:20)
    at TLSSocket.EventEmitter.emit (domain.js:482:12)
    at emitErrorNT (internal/streams/destroy.js:92:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
    at GetAddrInfoReqWrap.onlookupall [as oncomplete] (dns.js:76:26)
[0-0]  Error:  Failed to create session.
getaddrinfo ENOTFOUND hub-cloud.browserstack.com
[0-0] 2020-05-08T15:03:28.624Z ERROR @wdio/runner: Error: Failed to create session.
getaddrinfo ENOTFOUND hub-cloud.browserstack.com
    at startWebDriverSession (C:\WS\webdriverio-browserstack-master_6nonfunctional\node_modules\webdriver\build\utils.js:45:11)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
[0-0] FAILED in undefined - C:\WS\webdriverio-browserstack-master_6nonfunctional\tests\specs\single_test.js

Spec Files:      0 passed, 1 failed, 1 total (100% completed) in 00:00:01

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! webdriverio-browserstack@1.0.0 single: `wdio conf/single.conf.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the webdriverio-browserstack@1.0.0 single script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\s6200b\AppData\Roaming\npm-cache\_logs\2020-05-08T15_03_29_031Z-debug.log
```

**Additional context**
Please note that I used the example application just to make sure that there was a functioning baseline (that could be upgraded) to compare with but the same error occurs for me if I set up a WDIO v6 test from scratch as well when using Browserstack.  
I haver also tried using the wdio-browserstack-service but this results in the same error.

**Comment by dimkin-eu** [comment]
aint you missing `services: [["browserstack"]],` in wdio.conf.js?

**Comment by th0car** [comment]
@dimkin-eu - Afaik I don't use (or have to use) any service to target browserstack as grid. At least I don't need that using identical conf but with WebdrivierIO v5 and the browserstack. Which service are you thinking of? 
(As mentioned above I have deliberately chosen to not use the @wdio/browserstack-service to exemplify the problem)

**Comment by dimkin-eu** [comment]
@th0car apart dependency you need to set service part in wdio.conf.js
https://webdriver.io/blog/2020/03/26/webdriverio-v6-released.html#service-configurations
not the right link about services, but still

**Comment by th0car** [comment]
OK - So to avoid focus on the service part I constructed an example myself instead using the WebdriverIO v5 Getting Started Guide with the @wdio/browserstack-service added. All test pass - Both using browserstack and browserstack-local binary. After this i upgrade WDIO to v6 and all tests immediately starts failing with the error in my original comment. Same environment, same test, same conf same browserstack-local version.

package.js for tests passing:
```
{
  "name": "wdio-basic-bs-service",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "@wdio/cli": "^5.22.4"
  },
  "devDependencies": {
    "@wdio/browserstack-service": "^5.22.1",
    "@wdio/local-runner": "^5.22.4",
    "@wdio/mocha-framework": "^5.18.7",
    "@wdio/spec-reporter": "^5.22.4",
    "@wdio/sync": "^5.20.1",
    "browserstack-local": "^1.4.5",
    "chromedriver": "^81.0.0",
    "wdio-chromedriver-service": "^5.0.2"
  }
}
```
package.js that causes Failed to create session. getaddrinfo ENOTFOUND hub-cloud.browserstack.com:
```
{
  "name": "wdio-basic-bs-service",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {},
  "devDependencies": {
    "@wdio/browserstack-service": "^6.1.10",
    "@wdio/cli": "^6.1.11",
    "@wdio/local-runner": "^6.1.11",
    "@wdio/mocha-framework": "^6.1.8",
    "@wdio/spec-reporter": "^6.1.9",
    "@wdio/sync": "^6.1.8",
    "browserstack-local": "^1.4.5",
    "chromedriver": "^81.0.0",
    "wdio-chromedriver-service": "^6.0.3"
  }
}
```

conf used (identical for v5 and v6) This is the simple version that is not using browserstack local binary:
```
exports.config = {
  user: process.env.BROWSERSTACK_USERNAME || "myUsername",
  key: process.env.BROWSERSTACK_ACCESS_KEY || "myKey",
  browserstackLocal: false,
  runner: "local",
  specs: ["./test/specs/**/nonlocal.js"],
  exclude: [],
  maxInstances: 10,
  capabilities: [
    {
      maxInstances: 5,
      os_version: "10",
      browserName: "Chrome",
      "browserstack.local": "false",
      "browserstack.user": "myUser",
      "browserstack.key": "myKey",
      forceLocal: "false",
      proxyHost: "corporate.proxy.com",
      proxyPort: "8080",
    },
  ],
  logLevel: "info",
  bail: 1,
  baseUrl: "",
  waitforTimeout: 10000,
  connectionRetryTimeout: 90000,
  connectionRetryCount: 1,
  services: [
    //['chromedriver'],
    [
      "browserstack",
      {
        key: "myKey",
        browserstackLocal: false,
        proxyHost: "corporate.proxy.com",
        proxyPort: "8080",
        verbose: true,
      },
    ],
  ],
  framework: "mocha",
  reporters: ["spec"],
  mochaOpts: {
    ui: "bdd",
    timeout: 60000,
  },
};
```

**Comment by christian-bromann** [comment]
It seems like you are not connected to the internet or you are behind a firewall. The error `ENOTFOUND hub-cloud.browserstack.com` says that it can't find the page which actually exists.

**Comment by th0car** [comment]
Sorry to see this closed with this motivation. I do, of course, understand ENOTFOUND and if I would have had experienced this problem with other versions of WebdriverIO or any other node application I would certainly not have filed this as a bug.
 
But since I am using the same computer, same env variables, same .npmrc, same proxy and the same internet connection (and the same tests and the same wdio.config) for WebdriverIO v5 where this problem does not occur I still suspect that the problem is caused by WebdriverIO v6 or rather one of its dependencies. 
I'd be happy to help troubleshooting, but I need some help in trying to track down the root cause.

**Comment by christian-bromann** [comment]
Sorry for closing it too early. We have changed the underlying request library and a regression in this area is possible. Can you try to recreate the issue with a much smaller example, e.g. using WebdriverIO in a simple node script?`

**Comment by th0car** [comment]
After installing global-agent and setting the additional env variables for this package I was able to access hub-cloud.browserstack.com also from WebdriverIO v6. 
This proxy setup is accordning to the webdriverIO documentation for v6 and I should have looked there even though v5 required no extra proxy config at all. Closing.

**Comment by sknight80** [comment]
Hi @th0car ,

Are you saying you are running a selenium-server to connect to Browserstack?  We have the same problem that you mentioned at the beginning; the differences are that we are not using the proxy. What should we do to solve this problem and make the documentation up-to-date since Browserstack documentation is out-dated, and here the examples are not working with V6?  I am a little bit frustrating to keep up-to-date WDIO between version updates...

**Comment by th0car** [comment]
@sknight80 My company was participating in a Browserstack POC where I focused on getting our current WebdriverIO 5 implementation to execute tests targeting internal DEV-environments using Browserstack Local from a TFS pipeline. We succeeded in this after a couple of days involving a lot of trial and error and fine tuning windows environment variables.
 (btw - the only Selenium server involved is the one on the Browserstack side so that was never an issue). 

We also decided to test if WDIO 6 would work. For this test we stopped at just getting a simple test to run targeting our production environmen which didn't require the browserstack local binary. It was for this test that the problem in this issue occured. 
My mistake was to upgrade WDIO 5 to WDIO 6 **without** adding global-agent package and the associated env variables (see https://webdriver.io/docs/proxy.html). The reason I didn't think of doing this was that no additional proxy settings or proxy handling packages were required for the same setup in WDIO 5 before upgrading.  

So...  problem is solved and WDIO 6 documentation is correct as far as the proxy goes. The settings in my WDIO 6 package.json and *.conf above should be enough if you also add the settings described in the v 6 proxy docs. If you also need to run the browserstack local binary with WDIO 6 I'm afraid I don't have a working example yet.
